process: add execve

ShogunPanda opened the pull request:
This PR adds a new `process.execve` method (absolutely willing to change the name if you want), which is a wrapper for the `execve` UNIX function.

The function will never return and will swap the current process with a new one.
All memory and system resources are automatically collected from `execve`, except for std{in,out,err}.

The primary use of this function is in shell scripts to allow to setup proper logics and then spawn another command.

ljharb commented:
What does it do on non-Unix systems?

ShogunPanda commented:
@ljharb By non Unix we only mean Windows, isn't it? If that's the case, Windows also supports `execve` via `_execve` ([doc here](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execve-wexecve?view=msvc-170)) so we should be good to go.

Once the CI ends I'll see which platform needs special assistance.

targos commented:
On Windows:

```
D:\a\node\node\src\node_process_methods.cc(544,26): error C2065: 'F_GETFD': undeclared identifier [D:\a\node\node\libnode.vcxproj]
  (compiling source file '/src/node_process_methods.cc')
  
D:\a\node\node\src\node_process_methods.cc(544,17): error C3861: 'fcntl': identifier not found [D:\a\node\node\libnode.vcxproj]
  (compiling source file '/src/node_process_methods.cc')
  
D:\a\node\node\src\node_process_methods.cc(546,17): error C2065: 'FD_CLOEXEC': undeclared identifier [D:\a\node\node\libnode.vcxproj]
  (compiling source file '/src/node_process_methods.cc')
  
D:\a\node\node\src\node_process_methods.cc(547,16): error C2065: 'F_SETFD': undeclared identifier [D:\a\node\node\libnode.vcxproj]
  (compiling source file '/src/node_process_methods.cc')
  
D:\a\node\node\src\node_process_methods.cc(547,7): error C3861: 'fcntl': identifier not found [D:\a\node\node\libnode.vcxproj]
  (compiling source file '/src/node_process_methods.cc')
```

targos commented on test/parallel/test-process-replace.js:
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+require('../common');
```
```suggestion
const common = require('../common');
```

ShogunPanda replied:
Thanks, implemented locally!

targos commented on test/parallel/test-process-replace.js:
```diff
@@ -0,0 +1,21 @@
+
+require('../common');
+
+const { deepStrictEqual, fail, strictEqual } = require('assert');
```
```suggestion
const { deepStrictEqual, fail, strictEqual } = require('assert');

if (!common.isMainThread)
  common.skip('process.replace is not available in Workers');
```

ShogunPanda replied:
Thanks, implemented locally!

targos commented on test/parallel/test-process-replace-fail.js:
```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const { mustCall } = require('../common');
```
```suggestion
const { isMainThread, mustCall } = require('../common');
```

ShogunPanda replied:
Thanks, implemented locally!

targos commented on test/parallel/test-process-replace-fail.js:
```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const { mustCall } = require('../common');
+const { throws } = require('assert');
```
```suggestion
const { throws } = require('assert');

if (!common.isMainThread)
  common.skip('process.replace is not available in Workers');
```

targos commented on test/parallel/test-process-replace-socket.js:
```diff
@@ -0,0 +1,41 @@
+
+const { mustCall, mustNotCall } = require('../common');
+const { fail, ok } = require('assert');
+const { createServer } = require('net');
```
```suggestion
const { createServer } = require('net');

if (!common.isMainThread)
  common.skip('process.replace is not available in Workers');
```

ShogunPanda replied:
Thanks, implemented locally!

targos commented on test/parallel/test-process-replace-socket.js:
```diff
@@ -0,0 +1,41 @@
+'use strict';
+
+const { mustCall, mustNotCall } = require('../common');
```
```suggestion
const { isMainThread, mustCall, mustNotCall } = require('../common');
```

ShogunPanda replied:
Thanks, implemented locally!

targos reviewed (commented):
To fix https://ci.nodejs.org/job/node-test-commit-custom-suites-freestyle/40120/

ardinugrxha commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,102 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  for (int i = 0; i < 3; i ++) {
+    int flags = fcntl(i, F_GETFD, 0);
+    if (flags >= 0) {
+      flags &= ~FD_CLOEXEC;
```
there is no err check if the `fcntl` function fails ?

ShogunPanda replied:
Good spot. Added.

ardinugrxha commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,102 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    node::Utf8Value pathname_string(env->isolate(), args[0].As<String>());
+
+    argv = new char*[2];
+    argv[0] = strdup(*pathname_string);
```
we need check not null also in here, maybe(?)

ShogunPanda replied:
Good spot. Added.

anonrig replied:
I think we should use std::memcpy or a similar modern C++ function other than this.

mcollina reviewed (approved):
lgtm

richardlau commented on test/parallel/test-process-replace-fail.js:
```diff
@@ -0,0 +1,84 @@
+    process.replace(
+      process.execPath + '_non_existing',
+      [__filename, 'replaced'],
+      { EXECVE_A: 'FIRST', EXECVE_B: 'SECOND', CWD: process.cwd() }
```
You will need to extend, rather than replace, the environment to fix the sharedlib builds.

ShogunPanda replied:
Thanks for the hint! I'll update the test to include `process.env`.

ShogunPanda commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,102 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    free(envp[i]);
+  }
+
+  delete argv;
```
I updated it. By using `delete[]` also `delete` on each element got called so I removed the for cycles. Is that correct?

ardinugrxha replied:
yes it is

ShogunPanda commented:
@ljharb I was totally wrong. Despite of naming and similar signatures, the function `_execve` only creates a new process without replacing the old one.

I'll disable this function on Windows.

addaleax commented on doc/api/errors.md:
```diff
@@ -2445,6 +2445,19 @@ An invalid timestamp value was provided for a performance mark or measure.
 
+<a id="ERR_PLATFORM_UNSUPPORTED_OPERATION"></a>
+
+### `ERR_PLATFORM_UNSUPPORTED_OPERATION`
```
How does this differ from `ERR_FEATURE_UNAVAILABLE_ON_PLATFORM`?

ShogunPanda replied:
Simply I didn't find it. ;)
Removing it and using the existing one.

addaleax commented on doc/api/process.md:
```diff
@@ -3303,6 +3303,30 @@ In custom builds from non-release versions of the source tree, only the
+resources from the current process are preserved, except for the standard input,
+standard output and standard error file descriptor.
+
+All other resources are discarded by system when the processes are swapped.
```
What if this isn't the a desirable behavior in a given situation?

ShogunPanda replied:
Can you please expand this? What do you mean?

addaleax replied:
If there's any point in adding execve(), it's the customizability that the method brings with it. Leaving file descriptors open  and/or redirecting them intentionally can be part of that -- just like you could in a bash script do `exec bash 0<&4 1>&4` to create a shell that reads and writes to e.g. a pre-opened socket or something along those lines.

(With a similar reasoning, you could also allow users to actually specify `argv[0]` with a value they control -- it doesn't _need_ to equal the filename that's being executed)

addaleax commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+
+  THROW_ERR_PROCESS_REPLACE_FAILED(env, error_code);
+}
+#endif
```
This is quite hard-to-follow C++ with lots of unnecessary explicit memory management that Node.js has been doing a lot of effort to move away from. I'd recommend looking a bit a how other parts of the Node.js code base handle strings and conversion between C++ and JS values.

anonrig replied:
+1

ShogunPanda replied:
I'm absolutely willing to. Since I'm not familiar with the C++ codebase, do you have any suggestion of places I can look to?

addaleax replied:
@ShogunPanda Well, mostly anywhere else works. As a general rule, you'll want to get rid of `new`, `new[]`, `char*`, `memcpy()` and `strdup()` as much as possible, and replace them with `std::vector`, `std::string`/`Utf8Value` as much as possible. 

(You won't be entirely able to avoid something like `std::vector<char*>` because `execve` expects `char**` arguments, but the `std::vector<char*>`'s entries could point to the entries of a `std::vector<std::string>` or `std::vector<Utf8Value>` rather than having to manage memory manually).

addaleax replied:
> So I have to use two vectors for `argv` and two for `envp`. Is this the right approach or am I still missing something?

It can be two vectors or a single one, that doesn't really matter much 🙂

ShogunPanda replied:
Lol. Makes sense.
But how do you go from `std::vector<Utf8Value>` to `char**` without a second array?
I'll probably won't change the code anyway, I'm just trying to learn.

addaleax reviewed (changes requested):
**If** this is to be added, I'd absolutely recommend referring to it by a standard name such as `execve()`, or otherwise something that makes it clear that it spawns a new process.

This requires integration with the permissions API or is otherwise an immediate security hole.

Overall I'd recommend not adding this though, unless there's a concrete reason to believe that it fills a significant gap that the existing `child_process` API doesn't cover.

jasnell commented:
> Why would we want to add a function that can't work on all tier 1 platforms?

Well, we already have a number of such apis... `process.getegid()` for instance. There are actually quite a few already on `process`.

jasnell commented:
I don't consider it to be ideal. I would have preferred a pattern like `process.posix.getegid()` similar to what we have with `path.posix` but these predate my involvement so they are what they are.

jasnell commented:
+1 on @addaleax's alternative name suggestion. I'm fine with adding this so long as the cleanup logic/expectations are clearly documented.

anonrig commented on lib/internal/process/per_thread.js:
```diff
@@ -223,6 +229,49 @@ function wrapProcessMethods(binding) {
+    }
+
+    // Construct the environment array.
+    const envArray = ObjectEntries(env || {}).map(({ 0: key, 1: value }) => `${key}=${value}`);
```
This needs ArrayPrototypeMap or we can remove ObjectEntries. I prefer removing primordials, but I'm just a voice in a desert :-))

ShogunPanda replied:
No you're not. We are at least two. Lol.
Thanks for spotting that, removed.

anonrig commented on src/node_errors.h:
```diff
@@ -13,6 +13,10 @@
 #include <sstream>
 
 namespace node {
+// This forward declaration is to have the method available in error messages.
```
```suggestion
// This forward declaration is required to have the method available in error messages.
```

ShogunPanda replied:
Fixed, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  char** target = nullptr;
+  int length = js_array->Length();
+
+  CHECK_LT(length, INT_MAX);
```
```suggestion
  CHECK_LT(length, INT_MAX);
  CHECK_BT(length, 0);
```

ShogunPanda replied:
I couldn't find `CHECK_BT` anywhere. Did you mean `CHECK_GT` or what?

anonrig replied:
Yes!

ShogunPanda replied:
The function has been removed, so no need to this change anymore.

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    for (int i = 0; i < length; i++) {
+      node::Utf8Value str(env->isolate(),
+                          js_array->Get(context, i).ToLocalChecked());
+      target[i] = strdup(*str);
```
you can just use std::memcpy and write modern C++. strdup is a C API.

ShogunPanda replied:
Fixed, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    }
+  }
+
+  if (target_length != nullptr) {
```
Nit: Also works

```suggestion
  if (target_length) {
```

ShogunPanda replied:
Fixed, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  return target;
+}
+
+#if !defined(_WIN32)
```
```suggestion
#ifndef _WIN32
```

ShogunPanda replied:
Fixed, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    Local<Array> full_argv_array =
+        Array::New(env->isolate(), argv_array->Length() + 1);
+    full_argv_array->Set(context, 0, args[0].As<String>()).Check();
+    for (unsigned int i = 0; i < argv_array->Length(); i++) {
+      full_argv_array
+          ->Set(context, i + 1, argv_array->Get(context, i).ToLocalChecked())
+          .Check();
+    }
```
Does this make a copy? I didn't quite understand the goal here. Sorry for the questions!

ShogunPanda replied:
Yes, that's the idea. `execve` wants a null-terminated `char**` while I have a `JsArray`.
On the Javascript side I have validated they are all strings with no null-bytes in the middle.
Any better idea on how to handle this?

addaleax replied:
My gut feeling would be that it's much easier to handle the conditional and array copying on the JS side of things and instead only pass a version of the array(s) down to C++ that's close to what the syscall expects 🙂

ShogunPanda replied:
I see what you mean now.
I would love to make all the manipulation on the JS side, but it is then possible to pass the data from JS to be interpreted on C++ as `char**`. Out of my mind I'm thinking about using `Uint8Array` but I might be wrong.

addaleax replied:
You can still pass an array of strings, it's just the array copying/argument handling that would end up simplified (i.e. no need to check `if (args.Length() > 1) {`, no need to `full_argv_array->Set(context, i + 1, argv_array->Get(context, i).ToLocalChecked())`)

ShogunPanda replied:
This function has been removed.

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+
+    argv = copy_js_strings_array(env, full_argv_array, nullptr);
+  } else {
+    node::Utf8Value pathname_string(env->isolate(), args[0].As<String>());
```
Nit: node prefix is not needed

```suggestion
    Utf8Value pathname_string(env->isolate(), args[0].As<String>());
```

ShogunPanda replied:
Fixed, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+    if (argv[0] == nullptr) {
+      int error_code = errno;
+      delete[] argv;
+      THROW_ERR_PROCESS_REPLACE_FAILED(env, error_code);
```
return is missing after this line

ShogunPanda replied:
Thanks, added it!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -662,6 +786,10 @@ static void CreatePerIsolateProperties(IsolateData* isolate_data,
   SetMethod(isolate, target, "dlopen", binding::DLOpen);
   SetMethod(isolate, target, "reallyExit", ReallyExit);
+
+#if !defined(_WIN32)
```
```suggestion
#ifndef _WIN32
```

ShogunPanda replied:
Thanks, fixed it!

ShogunPanda commented on src/node_errors.h:
```diff
@@ -253,6 +258,13 @@ inline v8::Local<v8::Object> ERR_STRING_TOO_LONG(v8::Isolate* isolate) {
 }
 
+inline void THROW_ERR_PROCESS_REPLACE_FAILED(Environment* env, int code) {
+  std::ostringstream message;
```
I switch to `snprintf`. Not sure how I got into `ostringstream` :)

ShogunPanda commented:
> > Why would we want to add a function that can't work on all tier 1 platforms?
> 
> Well, we already have a number of such apis... `process.getegid()` for instance. There are actually quite a few already on `process`.


Thanks for involuntary hint. :)
I updated the code with `#ifdef __POSIX__` and the docs to use the same wording already used there.

ShogunPanda commented:
@addaleax 

> **If** this is to be added, I'd absolutely recommend referring to it by a standard name such as `execve()`, or otherwise something that makes it clear that it spawns a new process.
> 

As requested, I've renamed it to `execve`. I didn't like the original name either :)

> This requires integration with the permissions API or is otherwise an immediate security hole.

I just added integration with the permission API. It will require `--allow-child-process`.
 
> Overall I'd recommend not adding this though, unless there's a concrete reason to believe that it fills a significant gap that the existing `child_process` API doesn't cover.

In the shell scripting context, there is no way to create a new process which would replace the current one. Think about using Node.js to build a complex command to run. After it, Node.js was not used anymore but since there was no way to swap the process, you would have to spawn a new process, manage it's stdin/stdout/stderr and so forth. That's why I added `execve`.

ShogunPanda commented:
@jasnell 

> +1 on @addaleax's alternative name suggestion. I'm fine with adding this so long as the cleanup logic/expectations are clearly documented.

I added two tests which check that:

1. Test what happens if we leave a socket open and we try to reopen in the swapped process: the operation succeeds, which means the fd is destroyed. I also tried to install a `on('close')` but it is not invoked.

2. Test what happens if there is a `process.on('exit')` on the original process. It does not get invoked, which means that the system call immediately swap the programs without running any cleanup logic.

I've added this to the documentation to reflect that. Does it look good now?

ShogunPanda commented on doc/api/process.md:
```diff
@@ -3303,6 +3303,30 @@ In custom builds from non-release versions of the source tree, only the
+
+All other resources are discarded by system when the processes are swapped.
+
+This function will never return, unless an error occurred.
```
See my comment below.

addaleax commented:
> After it, Node.js was not used anymore but since there was no way to swap the process, you would have to spawn a new process, manage it's stdin/stdout/stderr and so forth. That's why I added `execve`.

I mean, to be clear, this is still a one-line operation (managing stdio literally just comes down to setting `stdio: 'inherit'`) for the most part. I know where you're coming from but I wouldn't add this just for the sake of adding it.

ShogunPanda commented:
> > After it, Node.js was not used anymore but since there was no way to swap the process, you would have to spawn a new process, manage it's stdin/stdout/stderr and so forth. That's why I added `execve`.
> 
> I mean, to be clear, this is still a one-line operation (managing stdio literally just comes down to setting `stdio: 'inherit'`) for the most part. I know where you're coming from but I wouldn't add this just for the sake of adding it.

I kinda agree.
You still would have to handle the exit code to ensure proper broadcasting to the shell.
Also, I'm not sure what would happen with "TUI" (ncurses and similar) programs and so forth.

Is your objection a full block or just an "intent"?

addaleax commented:
@ShogunPanda Just fyi, have you seen https://www.npmjs.com/package/foreground-child? That also works on Windows 🙂 

My "request changes" marker only refers to the C++ memory management here, i.e. [this comment](https://github.com/nodejs/node/pull/56496#discussion_r1905810009) specifically. I don't intend to block this feature, I've given my opinion but it's pretty clear overall that the Node.js project has a different stance from my own when it comes to 'what should go into core' 🙂

anonrig commented on src/node_errors.h:
```diff
@@ -253,6 +259,15 @@ inline v8::Local<v8::Object> ERR_STRING_TOO_LONG(v8::Isolate* isolate) {
 }
 
+inline void THROW_ERR_PROCESS_EXECVE_FAILED(Environment* env, int code) {
+  char message[128];
+  snprintf(message,
+           sizeof(message),
+           "process.execve failed with error code %s",
+           errors::errno_string(code));
```
Why not?

```suggestion
  auto message = std::string("process.execve failed with error code " + errors::errno_string(code);
```

ShogunPanda replied:
Updated, thanks!

anonrig commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,125 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+        Array::New(env->isolate(), argv_array->Length() + 1);
+    full_argv_array->Set(context, 0, args[0].As<String>()).Check();
+    for (unsigned int i = 0; i < argv_array->Length(); i++) {
+      full_argv_array
+          ->Set(context, i + 1, argv_array->Get(context, i).ToLocalChecked())
+          .Check();
```
If you don't want this to throw, you can surround this with USE() and remove .Check()

addaleax replied:
?

Ideally, this should use neither of the two and just do proper error handling

jasnell replied:
Indeed. this should probably be something like...

```c++
  if (full_argv_array->Set(...).IsNothing()) {
    return;
  }
```

Likewise the `Check()` and `ToLocalChecked()` uses immediately above should be changed to properly handle the error cases without resorting to abort.

This API call should only return nothing if an error is scheduled and returning immediately in that case allows v8 to proceed with throwing that error.

ShogunPanda replied:
I totally removed this function, no need for this anymore.

jasnell commented on lib/internal/process/per_thread.js:
```diff
@@ -223,6 +229,49 @@ function wrapProcessMethods(binding) {
+    if (!isMainThread) {
+      throw new ERR_WORKER_UNSUPPORTED_OPERATION('Calling process.execve');
+    } else if (process.platform === 'win32') {
+      throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('process.execve');
```
the docs say this is also unavailable on Android. Should that be checked here also?

targos replied:
I would remove Android from the docs. It's not a supported platform.

ShogunPanda replied:
Done, thanks!

jasnell commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,125 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  int error_code = errno;
+
+  // Free up memory, then throw.
+  delete[] argv;
```
The abundance of manual memory management in this function really makes me nervous.

jasnell commented on test/parallel/test-process-execve-fail.js:
```diff
@@ -0,0 +1,97 @@
+  skip('process.execve is not available in Workers');
+}
+
+if (!isWindows) {
```
The docs also mention Android as not being supported. Should that be covered here also?

ShogunPanda replied:
No, @targos was right. I have updated the docs.

jasnell commented on test/parallel/test-process-execve-fail.js:
```diff
@@ -0,0 +1,97 @@
+if (!isWindows) {
+  // Non existent path name
+  {
+    throws(mustCall(() => {
```
Passing a must call into `throws` doesn't make a lot of sense in my opinion. We know that `throws` is goingt to call that function immediately (we have tests proving this).

ShogunPanda replied:
Thanks, I removed that!

ShogunPanda commented:
@anonrig @addaleax @jasnell I have vastly refactored the C++ to address Anna's and James' suggestions. Can you please review it again?

Apparently I'm more a C guy than a C++ guy. And last time I officially touched C++ was several years ago. Lol.

Thanks for the patience here!

jasnell commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,76 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  argv[argv_array->Length()] = nullptr;
+
+  for (unsigned int i = 0; i < envp_array->Length(); i++) {
+    envp_strings[i] =
+        Utf8Value(isolate, envp_array->Get(context, i).ToLocalChecked())
+            .ToString();
```
For things like this, instead of using `ToLocalChecked()`, use a pattern like the following... this will avoid a process crash when a throw is appropriate:

```
Local<Value> str;
if (!envp_array->Get(context, i).ToLocal(&str)) {
  return;
}
envp_strings[i] = Utf8Value(str).ToString();
```

ShogunPanda replied:
Done, thanks!

jasnell commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,76 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  execve(*executable, argv.data(), envp.data());
+  int error_code = errno;
+
+  THROW_ERR_PROCESS_EXECVE_FAILED(env, error_code);
```
Given that you're running the `RunAtExit(...)` here, and given that the expectation is for this process to be replaced, I wonder if it would be more appropriate to crash the process here instead of just throwing.

ShogunPanda replied:
Good point.
I update that. Can you check if I used the proper methods to print the errors before aborting or was it there a better way?

Karinza38 reviewed (commented):
Karinza38

ShogunPanda commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,76 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+
+  // Set stdin, stdout and stderr to be non-close-on-exec
+  // so that the new process will inherit it.
+  for (int i = 0; i < 3; i++) {
```
Done, thanks!

addaleax commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,94 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  // so that the new process will inherit it.
+  if (persist_standard_stream(0) < 0 || persist_standard_stream(1) < 0 ||
+      persist_standard_stream(2) < 0) {
+    env->ThrowErrnoException(errno, "fcntl");
```
```suggestion
    env->ThrowErrnoException(errno, "fcntl");
    return;
```

addaleax commented on src/node_process_methods.cc:
```diff
@@ -471,6 +473,94 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+  std::string info = FormatErrorMessage(
+      isolate, context, error_message.c_str(), message, true);
+  FPrintF(stderr, "%s\n", info);
+  ABORT();
```
Why does this abort instead of throwing?

jasnell replied:
@addaleax ... see https://github.com/nodejs/node/pull/56496#discussion_r1912061490

I think this is something we can reasonably debate a bit.

addaleax replied:
Hm yeah ... I'd argue that this further limits the already fairly limited usefulness of this feature, but I understand where it's coming from 🙂 Just seems odd that this hard crashes in this case, you can't always know for sure that `execve()` will succeed

jasnell replied:
Yep, this one is rather tricky. I'm not quite sure what this should do really and absolutely open to alternatives. One suggestion is to mark this new API as experimental, stick with the abort for now, and see if a reasonable alternative emerges before we graduate it but I'm not really certain what the ideal approach here is.

ShogunPanda replied:
Done, the API is now experimental, we can revisit this in the future.

ShogunPanda commented:
@addaleax Are you still blocking this or are we good to merge?

ShogunPanda commented:
@nodejs/tsc No reply from @addaleax after two weeks. What shall I do?

ShogunPanda commented:
@aduh95 Thanks for the review, I'll take care of them soon. Since I consider you the "master of procedures and documentations", what is the procedure for waiting for objections re-evaluation?

aduh95 commented:
Our guidelines say https://github.com/nodejs/node/blob/96457b433fb1b9cdcc7d957914ae815aa40a5c78/doc/contributing/collaborator-guide.md?plain=1#L162-L166

I don’t know whether this block qualifies as "unclear to others" though, it might be one that would require a TSC vote

RafaelGSS commented on test/parallel/test-process-execve-permission-fail.js:
```diff
@@ -0,0 +1,21 @@
+} else {
+  throws(mustCall(() => {
+    process.execve(process.execPath, [process.execPath, __filename, 'replaced'], process.env);
+  }), { name: 'Error', code: 'ERR_ACCESS_DENIED' });
```
We should have a `permission: 'ChildProcess'` check here. No need to check `name`.

ShogunPanda replied:
Will do very soon. Thank!

ShogunPanda replied:
Added, thanks!

RafaelGSS reviewed (commented):
We should update the `--allow-child-process` to mention this new API as well.

ShogunPanda commented:
> Our guidelines say
> 
> https://github.com/nodejs/node/blob/96457b433fb1b9cdcc7d957914ae815aa40a5c78/doc/contributing/collaborator-guide.md?plain=1#L162-L166
> 
> I don’t know whether this block qualifies as "unclear to others" though, it might be one that would require a TSC vote

Her block was totally clear and I actually addressed her requests. But I haven't heard from her in two weeks after performing modifications. 

I'll bring this to the next TSC meeting to request unblocking.

ShogunPanda commented on lib/internal/process/per_thread.js:
```diff
@@ -223,6 +229,49 @@ function wrapProcessMethods(binding) {
+      }
+    }
+
+    if (env !== undefined) {
+      validateObject(env, 'env');
+
+      for (const { 0: key, 1: value } of ObjectEntries(env)) {
+        if (
+          typeof key !== 'string' ||
+          typeof value !== 'string' ||
+          StringPrototypeIncludes(key, '\u0000') ||
+          StringPrototypeIncludes(value, '\u0000')
+        ) {
+          throw new ERR_INVALID_ARG_VALUE(
+            'env', env, 'must be an object with string keys and values without null bytes',
+          );
+        }
+      }
+    }
+
+    // Construct the environment array.
+    const envArray = ArrayPrototypeMap(ObjectEntries(env || {}), ({ 0: key, 1: value }) => `${key}=${value}`);
```
Yes, we could. And it was actually much easier to read. Will be in the next push.

aduh95 commented on doc/api/process.md:
```diff
@@ -2516,8 +2516,7 @@ if (process.getuid) {
 
-This function is only available on POSIX platforms (i.e. not Windows or
-Android).
+This function not available on Windows.
```
Is this change related to this PR? Could this land separately, and should we update the YAML history section?

ShogunPanda replied:
It was already like that but docs were not updated so I did it here.

aduh95 commented on doc/api/process.md:
```diff
@@ -3343,6 +3342,33 @@ In custom builds from non-release versions of the source tree, only the
+
+Replaces the current process with a new process.
+
+This is achieved by using the `execve` Unix function and therefore no memory or other
```
Is it Unix or POSIX? This makes it sounds like it would be Unix-specific thing, but we say it’s available on POSIX below

ShogunPanda replied:
It was meant to be POSIX, updated!

RafaelGSS commented on lib/internal/process/per_thread.js:
```diff
@@ -269,6 +275,49 @@ function wrapProcessMethods(binding) {
   }
 
+  function execve(execPath, args, env) {
+    const { isMainThread } = require('internal/worker');
```
Since this is an experimental feature, we should emit a warning on usage.

ShogunPanda replied:
Added, thanks!

jasnell commented on doc/api/process.md:
```diff
@@ -3343,6 +3342,33 @@ In custom builds from non-release versions of the source tree, only the
+
+> Stability: 1 - Experimental
+
+* `file` {string} The name or path of the executable file to run.
```
Nit... not in this PR but as a separate follow on we should likely align this with the fs module handling of paths so that `file` can be represented using a `Buffer` or `URL` also.

ShogunPanda commented:
@addaleax The TSC has voted to dismiss your objection after waiting for a reply for two weeks. Thanks for your feedback anyway. ❤️

See: https://github.com/nodejs/TSC/pull/1705.

aduh95 commented:
@ShogunPanda thats not a vote, it’s a motion. You should have dismissed the objection using GH UI and land the PR using the CQ (or manually, but only after the objection was dismissed).

ShogunPanda commented:
Thanks @aduh95, I keep forgetting about the difference. :)
I amended the comment above.
I tried via the UI but couldn't find how to do that. Can you help me on that?
